Support different sysctl output formats

kiike opened the pull request:
In the io.popen deprecation PR @McSinyx mentioned that different platforms use different ways of outputting the sysctl info. For instance, in OpenBSD we use "=" between key and value, FreeBSD uses ": " and GNU/Linux uses " = ". Well, I think the sysctl helper should support that and account the difference between systems.

Here's a PR that intends to add such support. It's pretty straightforward and Works For Me(TM) BUT it should be tested in other systems. I've no Linux or FreeBSD systems with Awesome WM around here so please try this PR and let me know if it works for you.

This snippet in `rc.lua` was used to test it:
```lua
helpers = require("vicious.helpers")
helpers.sysctl_async({ "hw.usermem",
		       "hw.physmem"},
		     function(ret)
			for k, v in pairs(ret)
			do print(k, v) end
		     end
)
```
Yielding this output:

![](https://0x0.st/zwT6.png)

Cheers,
Enric

McSinyx commented on helpers.lua:
```diff
@@ -244,8 +244,17 @@ function helpers.sysctl_async(path_table, parse)
     spawn.with_line_callback("sysctl " .. path, {
         stdout = function(line)
+	    local separators = {
+	       openbsd = "=",
```
Vicious uses 4 spaces for indentation.  I assume you uses emacs which somehow always mess up the indentation level somewhere around 3 or 4 spaces :stuck_out_tongue:

kiike replied:
Addressed in commit. Thanks! :)

McSinyx commented on helpers.lua:
```diff
@@ -244,8 +244,17 @@ function helpers.sysctl_async(path_table, parse)
         stdout = function(line)
+	    local separators = {
+	       openbsd = "=",
+	       linux = " = ",
```
On GNU/Linux, `sysctl` is usually in `/sbin` and requires superuser permission so I'm not sure about the practical use of it for widgets.

McSinyx replied:
> I could use `sysctl` just fine on my Raspberry Pi although the output is very different than the BSD's. I don't know what use it could have either, but it doesn't hurt to have it I guess. :yum:

I obviously forgot that other than desktop computers, GNU/Linux also run on other hardware where permissions might be configured differently.  BTW, the error under GNU/Linux looks like this

    # sysctl foobar
    sysctl: cannot stat /proc/sys/foobar: No such file or directory

Please update the error handling accordingly along with the one on OpenBSD, preferably in the way you handled the separators.

kiike replied:
I've improved the error handling in a commit. Let me know if that's what you were expecting or what i can change.

McSinyx commented on helpers.lua:
```diff
@@ -244,8 +244,17 @@ function helpers.sysctl_async(path_table, parse)
+	    }
+
+	    local pattern = "(.+)%s(.+)"
+	    pattern = pattern:format(separators[helpers.getos()])
```
I personally believe this is unnecessarily verbose--if the pattern fit in the `match` below within 80 columns, please uses it anonymously.

kiike replied:
It was. I could've written it differently. Thanks.

McSinyx commented on helpers.lua:
```diff
@@ -244,8 +244,17 @@ function helpers.sysctl_async(path_table, parse)
+	    local pattern = "(.+)%s(.+)"
+	    pattern = pattern:format(separators[helpers.getos()])
+
             if not string.find(line, "sysctl: unknown oid") then
```
Since it is not shown in your screenshot, is `sysctl`'s error on OpenBSD has the same format?

kiike replied:
It doesn't, but that scenario is now taken care of. I could use help testing it on Linux, and FreeBSD though.

kiike commented:
Hi @mcsinyx! Thanks for having a look at it!
 
>Vicious uses 4 spaces for indentation. I assume you uses emacs which somehow always mess up the indentation level somewhere around 3 or 4 spaces 😛
 
Indeed. Emacs here with lua-mode. I'll have to check the spacing and reformat it. Good catch!
 
> On GNU/Linux, sysctl is usually in /sbin and requires superuser permission so I'm not sure about the practical use of it for widgets.
 
I could use sysctl just fine on my Raspberry Pi although the output is very different than the BSD's. I don't know what use it could have either, but it doesn't hurt to have it I guess.😋
 
> Since it is not shown in your screenshot, is  sysctl 's error on OpenBSD has the same format?
 
To be honest, I didn't account for that case, since I didn't have a clue as to what that error would be caused by. I'm out for the weekend but I'll look at it on Monday.
 
Cheers!

Sat Oct 05 04:17:42 GMT+02:00 2019 Nguyễn Gia Phong <notifications@github.com>:
 
> 
> @McSinyx  requested changes on this pull request.
>  
> Hi,
>  
> This looks really good to me, aside from a few styling issues (I'm writing a style guide along with the copyright fix to avoid manually telling contributors to re-format every time). BTW please add your name to the header (it is going to be fixed anyway in the next PR, but just to make sure I won't forget).
>  
> Thank you for making this contribution!
> 
>  
> In  helpers.lua [https://github.com/vicious-widgets/vicious/pull/93#discussion_r331728148]  :
>   > @@ -244,8 +244,17 @@ function helpers.sysctl_async(path_table, parse)
>  
>      spawn.with_line_callback("sysctl " .. path, {
>          stdout = function(line)
> +	    local separators = {
> +	       openbsd = "=",
>  
> Vicious uses 4 spaces for indentation. I assume you uses emacs which somehow always mess up the indentation level somewhere around 3 or 4 spaces  😛
>  
>  
> In  helpers.lua [https://github.com/vicious-widgets/vicious/pull/93#discussion_r331728233]  :
>   > @@ -244,8 +244,17 @@ function helpers.sysctl_async(path_table, parse)
>  
>      spawn.with_line_callback("sysctl " .. path, {
>          stdout = function(line)
> +	    local separators = {
> +	       openbsd = "=",
> +	       linux = " = ",
>  
> On GNU/Linux,  sysctl  is usually in  /sbin  and requires superuser permission so I'm not sure about the practical use of it for widgets.
>  
>  
> In  helpers.lua [https://github.com/vicious-widgets/vicious/pull/93#discussion_r331728400]  :
>   > @@ -244,8 +244,17 @@ function helpers.sysctl_async(path_table, parse)
>  
>      spawn.with_line_callback("sysctl " .. path, {
>          stdout = function(line)
> +	    local separators = {
> +	       openbsd = "=",
> +	       linux = " = ",
> +	       freebsd = ": "
> +	    }
> +
> +	    local pattern = "(.+)%s(.+)"
> +	    pattern = pattern:format(separators[helpers.getos()])
>  
> I personally believe this is unnecessarily verbose--if the pattern fit in the  match  below within 80 columns, please uses it anonymously.
>  
>  
> In  helpers.lua [https://github.com/vicious-widgets/vicious/pull/93#discussion_r331728455]  :
>   > @@ -244,8 +244,17 @@ function helpers.sysctl_async(path_table, parse)
>  
>      spawn.with_line_callback("sysctl " .. path, {
>          stdout = function(line)
> +	    local separators = {
> +	       openbsd = "=",
> +	       linux = " = ",
> +	       freebsd = ": "
> +	    }
> +
> +	    local pattern = "(.+)%s(.+)"
> +	    pattern = pattern:format(separators[helpers.getos()])
> +
>              if not string.find(line, "sysctl: unknown oid") then
>  
> Since it is not shown in your screenshot, is  sysctl 's error on OpenBSD has the same format?
>  
> —
> You are receiving this because you authored the thread.
> Reply to this email directly,  view it on GitHub [https://github.com/vicious-widgets/vicious/pull/93?email_source=notifications&email_token=AADRN4KRO7B62JCROABEBVDQM72MJA5CNFSM4I5VAG72YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCG7ZZRA#pullrequestreview-297770180]  , or  mute the thread [https://github.com/notifications/unsubscribe-auth/AADRN4MD5NNEWN5R4SCXZALQM72MJANCNFSM4I5VAG7Q]  . [https://github.com/notifications/beacon/AADRN4MA5W5YYKEWJM6ZK3DQM72MJA5CNFSM4I5VAG72YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCG7ZZRA.gif] 
>

kiike commented:
Hi @McSinyx,

I don't know how it works on my computer, but both on OpenBSD on my laptop and GNU/Linux on my Raspberry Pi both send errors to stderr. I reworked the parsing functions so that stderr can capture the different variations of errors and push "N/A" if the key is erroring.

I tried this snippet in a test `rc.lua`
```lua
helpers = require("vicious.helpers")
helpers.sysctl_async({"hw.usermem", "foo.bar", "hw.physmem"},
		     function(ret)
			for k, v in pairs(ret)
			do print(k, v) end
		     end)
```
 and I now get this, which looks nicer to me:
```
hw.usermem	8297222144
hw.physmem	8447561728
foo.bar	N/A
```
Instead of N/A, i could also write "ERROR" or the specific error, like "permission denied" or "unknown oid". I think N/A does the job though. On another note, shouldn't the sysctl command be spawned with an LANG=C (or equivalent) environment so that the output is not translated if the user has another language set?

I also addressed the incorrect indenting and the copyright line.

Let me know what you think

McSinyx commented:
> I don't know how it works on your computer, but both on OpenBSD on my laptop and GNU/Linux on my Raspberry Pi both send errors to stderr. 

It also prints to stderr on my system.  For FreeBSD, we'd need to wait for @mutlusun confirmation.  While on hold, you can try to refactor the battery widget for OpenBSD to use this new facility.

> I reworked the parsing functions so that stderr can capture the different variations of errors and push "N/A" if the key is erroring.

That's great.  Personally I think `N/A` is enough, since it indicates the field is missing.  On permission issues, I'm thinking about documenting it on a page about Vicious helpers which is currently lacking.

> On another note, shouldn't the sysctl command be spawned with an LANG=C (or equivalent) environment so that the output is not translated if the user has another language set?

Definitely! There's the existing `spawn.with_line_callback_with_shell` you can use for this.

mutlusun commented:
Hello,
thanks all for your work ;) I will test this PR this weekend on FreeBSD ;)

kiike commented:
> Hello,
> thanks all for your work ;) I will test this PR this weekend on FreeBSD ;)

Thanks in advance for trying it out. Looking forward to your feedback!

kiike commented:
> ... While on hold, you can try to refactor the battery widget for OpenBSD to use this new facility.

@McSinyx: done! It was a good way to further practice on async code, which has been above my skill level. I also took the chance to make the variables more readable.

kiike commented:
Thank you again for the feedback. I'll be out for the weekend so ill work on the styling on Monday. On your last question, sysctl outputs the value followed by the units and a description or comment. That rematching captures the value, discarding the unit and the description.
 
Have a great weekend too!

Sat Oct 12 09:11:36 GMT+02:00 2019 Nguyễn Gia Phong <notifications@github.com>:
 
> 
> @McSinyx  requested changes on this pull request.
>  
> There are still a few remaining styling issues. Please fix them accordingly and have a happy weekend!
> 
>  
> In  widgets/bat_openbsd.lua [https://github.com/vicious-widgets/vicious/pull/93#discussion_r333887324]  :
>   >  
> -return helpers.setasyncall{
> -    async = function (format, warg, callback)
> -        local filter = "hw.sensors.acpi" .. (warg or "bat0")
> -        local pattern = filter .. ".(%S+)=(%S+)"
> -        local bat_info = {}
> +    local sysctl_vars = {}
> +    for k, v in pairs(battery) do
> +        v = v:format(warg or "bat0")
> +        battery[k], sysctl_vars[#sysctl_vars+1] = v, v
>  
> There's  table.insert  if you find  foo[#foo] = bar  redundant. To me either is fine.
>  
>  
> In  widgets/bat_openbsd.lua [https://github.com/vicious-widgets/vicious/pull/93#discussion_r333890377]  :
>   >  
> -        spawn.with_line_callback_with_shell(
> -            ("sysctl -a | grep '^%s'"):format(filter),
> -            { stdout = function (line)
> -                  for key, value in line:gmatch(pattern) do
> -                      bat_info[key] = value
> -                  end
> -              end,
> -              output_done = function ()
> -                  -- current state
> -                  local state = STATES[tonumber(bat_info.raw0)]
> +    helpers.sysctl_async(sysctl_vars, function(ret)
> +    for k, v in pairs(battery) do
>  
> There is a missing indentation level here. Additionally anonymous function should be written as  function (ret)  instead, to make it consistent with  function foo(ret) .
>  
>  
> In  widgets/bat_openbsd.lua [https://github.com/vicious-widgets/vicious/pull/93#discussion_r333905592]  :
>   >  
> -                  local time
> -                  if tonumber(bat_info.power0) < 1 then
> -                      time = "∞"
> -                  else
> -                      local raw_time = bat_info.watthour3 / bat_info.power0
> -                      local hours, hour_fraction = math.modf(raw_time)
> -                      local minutes = math.floor(60 * hour_fraction)
> -                      time = ("%d:%0.2d"):format(hours, minutes)
> -                  end
> +    local charge = tonumber(battery.remaining_capacity
> +                                    / battery.last_full_capacity * 100)
>  
> The line continuation should align with the opening parenthesis.
>  
>  
> In  widgets/bat_openbsd.lua [https://github.com/vicious-widgets/vicious/pull/93#discussion_r333906272]  :
>   > -                  --  * battery state symbol (↯, -, !, + or N/A)
> -                  --  * remaining_capacity (in percent)
> -                  --  * remaining_time, by battery
> -                  --  * wear level (in percent)
> -                  --  * present_rate (in Watts)
> -                  callback{state, percent, time, wear, rate}
> -              end })
> -    end }
> +    -- Pass the following arguments to callback function:
> +    --  * battery state symbol (↯, -, !, + or N/A)
> +    --  * remaining capacity (in percent)
> +    --  * remaining time, as reported by the battery
> +    --  * wear level (in percent)
> +    --  * present_rate (in Watts/hour)
> +    return callback({state, charge, remaining_time,
> +                        wear, battery.charging_rate})
>  
> Same here. And if you don't mind, use the  { foo, bar }  style that Jörg recommends.
>  
>  
> In  helpers.lua [https://github.com/vicious-widgets/vicious/pull/93#discussion_r333907038]  :
>   > @@ -244,9 +245,29 @@ function helpers.sysctl_async(path_table, parse)
>  
>      spawn.with_line_callback("sysctl " .. path, {
>          stdout = function(line)
> -            if not string.find(line, "sysctl: unknown oid") then
> -                local key, value = string.match(line, "(.+): (.+)")
> -                ret[key] = value
> +	    local separators = {
>  
> Something is still wrong with the indentation within this function. BTW,  function (line)  is much more preferred.
>  
>  
> In  widgets/bat_openbsd.lua [https://github.com/vicious-widgets/vicious/pull/93#discussion_r334220283]  :
>   >  
> -        spawn.with_line_callback_with_shell(
> -            ("sysctl -a | grep '^%s'"):format(filter),
> -            { stdout = function (line)
> -                  for key, value in line:gmatch(pattern) do
> -                      bat_info[key] = value
> -                  end
> -              end,
> -              output_done = function ()
> -                  -- current state
> -                  local state = STATES[tonumber(bat_info.raw0)]
> +    helpers.sysctl_async(sysctl_vars, function(ret)
> +    for k, v in pairs(battery) do
> +        battery[k] = tonumber(ret[v]:match("(.-) "))
>  
> I'm particularly against variable recycling in this case. Something like this would be preferred
>   if
>  warg 
> ==
>  
> nil
>  
> then
>  warg 
> =
>  
> "
> bat0
> "
>  
> end
> 
> local
>  fields 
> =
>  {
>     charging_rate 
> =
>  (
> "
> hw.sensors.acpi%s.power0
> "
> ):
> format
> (warg),
>     last_full_capacity 
> =
>  (
> "
> hw.sensors.acpi%s.watthour0
> "
> ):
> format
> (warg),
>     remaining_capacity 
> =
>  (
> "
> hw.sensors.acpi%s.watthour3
> "
> ):
> format
> (warg),
>     design_capacity 
> =
>  (
> "
> hw.sensors.acpi%s.watthour4
> "
> ):
> format
> (warg),
>     state 
> =
>  (
> "
> hw.sensors.acpi%s.raw0
> "
> ):
> format
> (warg)
> }
> 
> local
>  sysctl_vars 
> =
>  {}
> for
>  k, v 
> in
>  
> pairs
> (fields) 
> do
>  
> table.insert
> (sysctl_vars, v) 
> end
> 
> 
> local
>  battery 
> =
>  {}
> helpers.
> sysctl_async
> (sysctl_vars, 
> function
>  (
> ret
> )
>     
> for
>  k, v 
> in
>  
> pairs
> (fields) 
> do
>  battery[k] 
> =
>  
> tonumber
> (ret[v]:
> match
> (
> "
> (.-) 
> "
> ) 
> end
> 
> end
> )
>  
> BTW could you please explain why it is necessary to re-match the value by  "(.-) " ?
>  
> —
> You are receiving this because you authored the thread.
> Reply to this email directly,  view it on GitHub [https://github.com/vicious-widgets/vicious/pull/93?email_source=notifications&email_token=AADRN4I4PIWI2OIDD5Z7TKLQOF2CLA5CNFSM4I5VAG72YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCHU7NXI#pullrequestreview-300545757]  , or  unsubscribe [https://github.com/notifications/unsubscribe-auth/AADRN4OYW5MS3J7NLDLJCVLQOF2CLANCNFSM4I5VAG7Q]  . [https://github.com/notifications/beacon/AADRN4MLJHGTIYK23PEGGIDQOF2CLA5CNFSM4I5VAG72YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCHU7NXI.gif] 
>

mutlusun commented on helpers.lua:
```diff
@@ -244,9 +245,29 @@ function helpers.sysctl_async(path_table, parse)
+		linux = " = ",
+		openbsd = "="
+	    }
+            local pattern = ("(.+)%s(.+)"):format(separators[helpers.getos()])
```
In our `helpers.lua` there are multiple locations where we call `getos()`. Maybe we could save the output once at startup in a global variable and remove this overhead? (`getos()` is not async) Though, I don't think, it's much overhead ...

McSinyx replied:
Actually the output is saved as `kernel_name` so `io.popen` should be called at most once with the module is initialized.  Deprecate the `getos` completely may reduce a bit of overhead (one call and one conditional) but I agree that it's negligible.

kiike commented:
You reviews are very thourough, @McSinyx. I am learning a lot about this project's coding style and I hope that my code can be soon on par with the project's.

McSinyx commented on widgets/bat_openbsd.lua:
```diff
@@ -19,67 +19,71 @@
+	    --  * wear level (in percent)
+	    --  * present_rate (in Watts/hour)
+	    return callback({ state, charge, remaining_time,
+			    wear, battery.charging_rate })
```
`wear` should align with `state` from the line above.

kiike replied:
done!

McSinyx commented on widgets/bat_openbsd.lua:
```diff
@@ -19,67 +19,71 @@
-                  local percent = tonumber(
-                      bat_info.watthour3 / bat_info.watthour0 * 100)
+    local battery = {}
+    helpers.sysctl_async(sysctl_args, function(ret)
```
Please add the space before the parameter list of the anonymous function like what you did with #92.

McSinyx commented on widgets/bat_openbsd.lua:
```diff
@@ -19,67 +19,71 @@
-                  -- current state
-                  local state = STATES[tonumber(bat_info.raw0)]
+    local sysctl_args = {}
+    for _, v in pairs(fields) do table.insert(sysctl_args, v) end
```
I believe you missed localizing `table.insert` here.

kiike replied:
done!

McSinyx commented on widgets/bat_openbsd.lua:
```diff
@@ -19,67 +19,71 @@
+	last_full_capacity = ("hw.sensors.acpi%s.watthour0"):format(battery_id),
+	remaining_capacity = ("hw.sensors.acpi%s.watthour3"):format(battery_id),
+	design_capacity = ("hw.sensors.acpi%s.watthour4"):format(battery_id),
+	state = ("hw.sensors.acpi%s.raw0"):format(battery_id)
```
We have some problems with the indentations here.

McSinyx commented on widgets/bat_openbsd.lua:
```diff
@@ -19,67 +19,71 @@
-                 [255] = "N/A" }    -- unimplemented by the driver
+local bat_openbsd = {}
+function bat_openbsd.async(format, warg, callback)
+    local battery_id = warg or "bat0"
```
Just realized that this is much better than overriding `warg` :wink:

kiike replied:
Awesome! :)

McSinyx commented on helpers.lua:
```diff
@@ -244,9 +245,29 @@ function helpers.sysctl_async(path_table, parse)
+                openbsd = {"level name .+ in (.+) is invalid"},
+                linux = {"cannot stat /proc/sys/(.+):",
+                         "permission denied on key '(.+)'"},
+                freebsd = {"unknown oid '(.+)'"}
```
Please insert the spaces inside the braces.

McSinyx commented on helpers.lua:
```diff
@@ -244,9 +245,29 @@ function helpers.sysctl_async(path_table, parse)
+            }
+
+            for i=1,#messages[helpers.getos()] do
+                local key = line:match(messages[helpers.getos()][i])
```
I find this a bit difficult to read.  Maybe `ipairs` could be a better idea?

kiike commented:
I wanted to test the changes in my main X11 session (i normally test then in Xephyr) so the emacs changes around whitespace and indent-mode were lost. Silly me! Well, one of my previous commits had mixed tabs and spaces. That should be taken care of now. I also addressed your other comments. I hope I haven't missed any.

McSinyx commented:
Answering myself:

> On GNU/Linux, sysctl is usually in /sbin and requires superuser permission
> so I'm not sure about the practical use of it for widgets.

Other widget types simply add /sbin to `PATH`:

https://github.com/vicious-widgets/vicious/blob/a542c5dbe20295eec47c4e75228bd30c3c13ff4b/widgets/wifi_linux.lua#L48

However, it's very likely that normal user doesn't have permission to read /proc/sys.

kiike commented on helpers.lua:
```diff
@@ -245,32 +245,32 @@ function helpers.sysctl_async(path_table, parse)
 
-            for i=1,#messages[helpers.getos()] do
-                local key = line:match(messages[helpers.getos()][i])
+            for _, error_message in ipairs(messages[helpers.getos()]) do
```
done! i am the one who should be sorry for not having checked it more thoroughly. This is the kind of stuff that no matter how many times I look at it, it always looks fine to me 😿

McSinyx replied:
That's a reason we human work together: improve each other's work!  BTW I've merged the PR, and please consider my proposal (you joining us as a maintainer).

kiike replied:
I'll ask @Mic92 right now! In the meantime, I am adapting `widgets.cpu_freebsd` to work on OpenBSD.

kiike commented on helpers.lua:
```diff
@@ -245,32 +245,32 @@ function helpers.sysctl_async(path_table, parse)
             end
         end,
-        output_done = function() parse(ret) end
+        output_done = function () parse(ret) end
```
Done!

McSinyx commented:
That's right.  I've just tested on my Fedora workstation and `sysctl` is even in user's default PATH (somehow /usr/sbin was decided to be added).  It's still able to display quite many options.

kiike commented:
@McSinyx IIRC most major distros merged all the */bin directories (/usr/local/bin, /usr/bin, etc/) into one, and Fedora was one of the first ones.

McSinyx commented:
I'm surprised that they merged sbin too.  Thanks for the info btw.
